packages/wescheme-blocks/src/ast.tsx

```diff
@@ -16,50 +17,79 @@ function pluralize(noun: string, set: unknown[]) {
   return set.length + " " + noun + (set.length != 1 ? "s" : "");
 }
 
+const WeSchemeAST = createASTSpec({
+  ...CoreAST.config,
+  letLikeExpr: {
+    spec: Spec.nodeSpec([
+      Spec.value<string, "form">("form"),
+      Spec.required("bindings"),
+      Spec.required("expr"),
+    ]),
+  },
+  whenUnlessExpr: {
+    spec: Spec.nodeSpec([
+      Spec.value<string, "form">("form"),
+      Spec.required("predicate"),
+      Spec.required("exprs"),
+    ]),
+  },
+});
+
 /* Override the default pretty printer for Sequences,
  * so that they print as s-expressions
  */
 export function Sequence<Expr extends ASTNode = ASTNode>(
   from: Pos,
   to: Pos,
   exprs: Expr[],
-  name: ASTNode,
+  name: ASTNode | null,
   options = {}
 ) {
-  return new ASTNode({
+  return WeSchemeAST.makeNode({
     from,
     to,
     type: "sequence",
     fields: { name, exprs },
     options,
     ...Nodes.SequenceProps,
-    pretty: (node) => P.standardSexpr(node.fields.name, node.fields.exprs),
+    pretty: (node) =>
+      P.standardSexpr(node.fields.name || "", node.fields.exprs),
   });
 }
 
 export function LetLikeExpr(
   from: Pos,
   to: Pos,
   form: string,
-  bindings: ASTNode<{ exprs: ASTNode<{ name: ASTNode; body: ASTNode }>[] }>,
+  bindings: ASTNode,
   expr: ASTNode,
   options = {}
 ) {
-  return new ASTNode({
+  return WeSchemeAST.makeNode({
     from,
     to,
     type: "letLikeExpr",
     fields: { form, bindings, expr },
     options,
     pretty(node) {
+      if (!WeSchemeAST.isNodeOfType(node.fields.bindings, "sequence")) {
+        throw new Error(
+          `Expected letLikeExpr to be constructed with sequence for its binding field`
+        );
+      }
       return P.lambdaLikeSexpr(
         node.fields.form,
         P.horz(
           "(",
           P.sepBy(
-            node.fields.bindings.fields.exprs.map((expr) =>
-              P.standardSexpr(expr.fields.name, [expr.fields.body])
-            ),
+            node.fields.bindings.fields.exprs.map((expr) => {
+              if (!WeSchemeAST.isNodeOfType(expr, "variableDefinition")) {
+                throw new Error(
+                  `Expected letLikeExpr to have sequence of variableDefinition nodes`
+                );
+              }
+              return P.standardSexpr(expr.fields.name, [expr.fields.body]);
+            }),
```

Comment:
Here is an example of doing runtime type checking to narrow down the type of grandchildren (and great grandchildren) in the pretty functions. This validation wasn't happening at all before. Now typescript forces us to do the validation at runtime because it doesn't have enough information to do it statically (it's not in the node spec!)